src/frequenz/sdk/data_ingestion/resampling/timeseries_resampling_actor.py

```diff
@@ -0,0 +1,249 @@
+        self._output_senders: Dict[TimeSeriesId, Sender[Sample]] = {}
+        # Looking at these three ^ I am thinking about creating a class that would
+        # manage a subscription per single TimeSeriesId and hold all the channels for
+        # that purpose, e.g. `SubscriptionHandler`:
+        #
+        # class SubscriptionHandler:
+        #     input_receiver: Receiver[Sample]
+        #     output_channel: Broadcast[Sample]
+        #     output_sender: Sender[Sample]
+        #
+        # and then we would have just a single attribute, e.g.
+        # self._subscription_handlers: Dict[TimeSeriesId, SubscriptionHandler] = {}
+        self._resampling_timer = Timer(interval=self._resampling_period_s)
+
+    async def subscribe(self, subscription: Subscription) -> Receiver[Sample]:
+        """Subscribe for data for a specific time series.
+
+        Args:
+            subscription: subscription consisting of `time_series_id`, which is made up
+                of a (component_id, metric_id) pair and `start_time`, which is the
+                point in time from which to subscribe  for data or `None` to subscribe
+                for data from the live component data stream
+
+        Returns:
+            receiver for receiving samples for a specific time series id
+        """
+        if subscription.time_series_id not in self._output_channels:
+            response: SubscriptionResponse = await self._data_provider.subscribe(
+                subscription=subscription,
+            )
+            self._input_receivers[subscription.time_series_id] = response.receiver
+            output_channel = Broadcast[Sample](str(subscription.time_series_id))
+            self._output_channels[subscription.time_series_id] = output_channel
+            self._output_senders[
+                subscription.time_series_id
+            ] = output_channel.get_sender()
+            self._resampler.add_time_series(time_series_id=subscription.time_series_id)
+
+        return self._output_channels[subscription.time_series_id].get_receiver()
+
+    def is_valid_sample(self, sample: Optional[Sample]) -> bool:
+        """Check if the provided sample is valid.
+
+        Args:
+            sample: sample to be validated
+
+        Returns:
+            True if the sample is valid, False otherwise
+        """
+        if sample is None or sample.value is None or math.isnan(sample.value):
+            return False
+        return True
+
+    async def run(self) -> None:
+        """Run the actor."""
+        while True:
+            select = Select(
+                resampling_timer=self._resampling_timer,
+                subscription_receiver=self._subscription_receiver,
+                component_data_receiver=MergeNamed(**self._input_receivers),
+            )
+            while await select.ready():
+                if msg := select.subscription_receiver:
+                    subscription_request: SubscriptionRequest = msg.inner
```

Comment:
You are forgetting to check for `None`. If `msg.inner` is `None`, then the underlaying channel was closed. You need to handle this case. This applies for all selectable receivers.

Reply:
Ok. Then I should probably update some of these cases in the `microgrid_data.py` too, because I didn't account for `None` there. 

Here, I will handle the subscription only if `msg.inner` is not `None`. I think that should be enough in this case? The actor would continue running, but wouldn't be accepting new subscriptions, because that channel would be closed.

Reply:
Yeah, I guess we missed that in the review, but `None` means the underlaying channel is closed. I would just raise an exception (which will lead to the actor to be killed and restarted). We will change the channels implementation soon so it raises an exception if the channel is closed.

Reply:
Added checks for `None` being sent through the channel.

Reply:
Added a check for `None` and a `ConnectionError` will be raised when the subscription channel gets closed.